File: examples/custom-python/after/prevent.py
"""
AFTER: same pipeline + AgentSonar with Prevent Mode (auto-stop).

Run:
    pip install agentsonar
    python prevent.py

What you'll see:
    - WARNING fires at rotation 5.
    - At rotation 10, PreventError raises and the loop stops.
    - The except block prints the cycle path and rotation count.
    - Mocked cost is roughly one third of detect.py's because the loop
      stopped before burning the rest of the rotations.

Compare this to detect.py: only difference is the prevent={...} config
and the try/except around the loop.
"""
from agentsonar import monitor_orchestrator, PreventError

INPUT_TOKENS_PER_CALL = 3_000
OUTPUT_TOKENS_PER_CALL = 1_500
INPUT_PRICE_PER_M = 3.00
OUTPUT_PRICE_PER_M = 15.00

sonar = monitor_orchestrator(config={
    # Stop the silent loop at exactly 10 rotations.
    "prevent": {"cyclic_delegation": {"max_rotations": 10}},
})

rotations_done = 0
try:
    while True:
        sonar.delegation(source="researcher", target="writer")
        sonar.delegation(source="writer", target="reviewer")
        sonar.delegation(source="reviewer", target="researcher")
        rotations_done += 1

except PreventError as e:
    print(f"Stopped: {e.reason}")
    print(f"Cycle:   {' -> '.join(e.cycle_path)}")
    print(f"After:   {e.rotations} rotations (severity {e.severity})")

sonar.shutdown()

calls = rotations_done * 3
input_cost = calls * INPUT_TOKENS_PER_CALL * INPUT_PRICE_PER_M / 1_000_000
output_cost = calls * OUTPUT_TOKENS_PER_CALL * OUTPUT_PRICE_PER_M / 1_000_000
total = input_cost + output_cost
print(f"{rotations_done} rotations completed before auto-stop.")
print(f"Mocked cost: ${total:,.2f}")
print("Saved versus the 'before' run: roughly $9.67 of the $14.50 baseline.")

File: examples/node/after/prevent.ts
/**
 * AFTER: same pipeline + AgentSonar with Prevent Mode (auto-stop).
 *
 * Run:
 *     npm install
 *     npm run prevent
 *
 * What you'll see:
 *     - WARNING fires at rotation 5.
 *     - At rotation 10, PreventError throws and the loop stops.
 *     - The catch block prints the cycle path and rotation count.
 *     - Mocked cost is roughly one third of detect.ts's because the loop
 *       stopped before burning the rest of the rotations.
 *
 * Compare this to detect.ts: only difference is the prevent={...} config
 * and the try/catch around the loop.
 */
import { AgentSonar, PreventError } from 'agentsonar'

const INPUT_TOKENS_PER_CALL = 3_000
const OUTPUT_TOKENS_PER_CALL = 1_500
const INPUT_PRICE_PER_M = 3.0
const OUTPUT_PRICE_PER_M = 15.0

const sonar = new AgentSonar(
  {
    // Stop the silent loop at exactly 10 rotations.
    prevent: { cyclicDelegation: { maxRotations: 10 } },
  },
  'researcher-writer-reviewer',
)

let rotationsDone = 0
try {
  while (true) {
    sonar.delegation('researcher', 'writer')
    sonar.delegation('writer', 'reviewer')
    sonar.delegation('reviewer', 'researcher')
    rotationsDone += 1
  }
} catch (err) {
  if (err instanceof PreventError) {
    console.log(`Stopped: ${err.reason}`)
    console.log(`Cycle:   ${err.cyclePath.join(' -> ')}`)
    console.log(`After:   ${err.rotations} rotations (severity ${err.severity})`)
  } else {
    throw err
  }
}

sonar.shutdown()

const calls = rotationsDone * 3
const inputCost = (calls * INPUT_TOKENS_PER_CALL * INPUT_PRICE_PER_M) / 1_000_000
const outputCost = (calls * OUTPUT_TOKENS_PER_CALL * OUTPUT_PRICE_PER_M) / 1_000_000
const total = inputCost + outputCost
console.log(`${rotationsDone} rotations completed before auto-stop.`)
console.log(`Mocked cost: $${total.toFixed(2)}`)
console.log("Saved versus the 'before' run: roughly $9.67 of the $14.50 baseline.")
